src/framework/audio/internal/audiobuffer.h

```diff
@@ -74,6 +76,9 @@ class AudioBuffer
     samples_t m_renderStep = 0;
 
     std::shared_ptr<IAudioSource> m_source = nullptr;
+
+    std::mutex m_readMutex;
```

Comment:
Unfortunately, this is not the solution we would like to see. Our audio buffer [is designed to be look-free](https://github.com/musescore/MuseScore/pull/14612) to ensure maximum performance. This change is a step backward for the application.

Reply:
I looked for lock-free ring buffers that support thread-safe resizing, but I was not able to find any (if you happen to know any, I am open to suggestions).

My implementation is still non-blocking during normal operation, because there are separate mutexes for both threads that access the buffer. The only time the mutexes block is when the buffer is resized because the audio device has changed. I also tried to make the locked sections as long as possible.
So the only overhead caused by these mutexes are the system calls to lock and unlock them and a stricter memory guarantee (?). From this [SO question](https://stackoverflow.com/questions/3652056/how-efficient-is-locking-and-unlocked-mutex-what-is-the-cost-of-a-mutex) I get that the two mutexes should be on different cache lines as well to optimize the performance. I did not profile the buffer before and after, but there seems to be some code to do that.

If you absolutely cannot live with this, the only other options are either:
1. always oversize the buffer for more channels
2. make sure the buffer size is a multiple of 6 and have 3x fewer samples buffered for surround audio compared to stereo

I think both of these are hacky and do not extend well to any other number of channels in the way resizing does.